File: lib/check_done.py
#!/usr/bin/env python3
"""
SPIRAL Phase C — Check Done
Exits 0 if: all prd.json stories pass AND latest test report has 0 failures.
Exits 1 otherwise (loop continues).
"""
import argparse
import json
import os
import sys
import time


def find_latest_report(reports_dir: str) -> str | None:
    if not os.path.isdir(reports_dir):
        return None
    subdirs = sorted(
        [d for d in os.listdir(reports_dir) if os.path.isdir(os.path.join(reports_dir, d))],
        reverse=True,
    )
    for d in subdirs:
        candidate = os.path.join(reports_dir, d, "report.json")
        if os.path.isfile(candidate):
            return candidate
    return None


def main() -> int:
    parser = argparse.ArgumentParser(description="SPIRAL check-done gate")
    parser.add_argument("--prd", default="prd.json", help="Path to prd.json")
    parser.add_argument("--reports-dir", default="test-reports", help="Test reports directory")
    args = parser.parse_args()

    # ── Check prd.json ────────────────────────────────────────────
    if not os.path.isfile(args.prd):
        print(f"[check_done] ERROR: {args.prd} not found", file=sys.stderr)
        return 1

    with open(args.prd, encoding="utf-8") as f:
        prd = json.load(f)

    stories = prd.get("userStories", [])
    total = len(stories)
    pending = [s for s in stories if not s.get("passes")]
    done = total - len(pending)

    print(f"[check_done] PRD: {done}/{total} stories complete, {len(pending)} pending")

    if pending:
        print("[check_done] Pending stories:")
        for s in pending:
            print(f"  [{s['id']}] {s['title']} (priority: {s.get('priority', '?')})")

    # ── Check latest test report ──────────────────────────────────
    report_path = find_latest_report(args.reports_dir)
    if not report_path:
        print(
            f"[check_done] WARNING: No test report found in {args.reports_dir}/ — run tests first",
            file=sys.stderr,
        )
        print("[check_done] RESULT: INCOMPLETE (no test report)")
        return 1

    report_age_min = (time.time() - os.path.getmtime(report_path)) / 60
    if report_age_min > 120:
        print(
            f"[check_done] WARNING: Test report is {report_age_min:.0f} min old — "
            "results may be stale; re-run tests for accurate check",
            file=sys.stderr,
        )

    with open(report_path, encoding="utf-8") as f:
        report = json.load(f)

    summary = report.get("summary", {})
    failed = summary.get("failed", 0)
    errored = summary.get("errored", 0)
    passed = summary.get("passed", 0)
    test_total = summary.get("total", 0)
    pass_rate = summary.get("pass_rate", "?")

    print(f"[check_done] Tests ({os.path.basename(os.path.dirname(report_path))}): "
          f"{passed}/{test_total} pass ({pass_rate}), {failed} failed, {errored} errored")

    # ── Decision ─────────────────────────────────────────────────
    prd_done = len(pending) == 0
    tests_clean = (failed == 0 and errored == 0)

    if prd_done and tests_clean:
        print("[check_done] RESULT: SPIRAL COMPLETE — all stories done and 100% tests pass!")
        return 0

    reasons = []
    if not prd_done:
        reasons.append(f"{len(pending)} pending stories")
    if not tests_clean:
        reasons.append(f"{failed + errored} test failure(s)")
    print(f"[check_done] RESULT: INCOMPLETE ({', '.join(reasons)})")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: lib/spiral-core/src/check_done.rs
/// check_done subcommand — ports check_done.py
use crate::prd;
use std::path::Path;

fn find_latest_report(reports_dir: &str) -> Option<String> {
    if !Path::new(reports_dir).is_dir() {
        return None;
    }
    let entries = std::fs::read_dir(reports_dir).ok()?;
    let mut subdirs: Vec<String> = entries
        .filter_map(|e| e.ok())
        .filter(|e| e.path().is_dir())
        .map(|e| e.file_name().to_string_lossy().to_string())
        .collect();
    subdirs.sort_by(|a, b| b.cmp(a)); // newest-first

    for d in subdirs {
        let candidate = format!("{}/{}/report.json", reports_dir, d);
        if Path::new(&candidate).exists() {
            return Some(candidate);
        }
    }
    None
}

pub fn run(prd_path: &str, reports_dir: &str) -> i32 {
    if !Path::new(prd_path).exists() {
        eprintln!("[check_done] ERROR: {} not found", prd_path);
        return 1;
    }

    let prd_val = match prd::read_json(prd_path) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("[check_done] ERROR: {}", e);
            return 1;
        }
    };

    let errors = prd::validate_prd(&prd_val);
    if !errors.is_empty() {
        eprintln!("[schema] PRD validation failed:");
        for e in &errors {
            eprintln!("  - {}", e);
        }
        return 1;
    }

    let stories = prd_val["userStories"].as_array().unwrap();
    let total = stories.len();
    let pending: Vec<&serde_json::Value> = stories
        .iter()
        .filter(|s| {
            !s.get("passes")
                .and_then(|v| v.as_bool())
                .unwrap_or(false)
                && !s
                    .get("_decomposed")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false)
        })
        .collect();
    let done = total - pending.len();

    println!(
        "[check_done] PRD: {}/{} stories complete, {} pending",
        done,
        total,
        pending.len()
    );

    if !pending.is_empty() {
        println!("[check_done] Pending stories:");
        for s in &pending {
            let id = s.get("id").and_then(|v| v.as_str()).unwrap_or("?");
            let title = s.get("title").and_then(|v| v.as_str()).unwrap_or("?");
            let priority = s.get("priority").and_then(|v| v.as_str()).unwrap_or("?");
            println!("  [{}] {} (priority: {})", id, title, priority);
        }
    }

    // ── Check latest test report ─────────────────────────────────────────────
    let report_path = match find_latest_report(reports_dir) {
        Some(p) => p,
        None => {
            eprintln!(
                "[check_done] WARNING: No test report found in {}/ — run tests first",
                reports_dir
            );
            println!("[check_done] RESULT: INCOMPLETE (no test report)");
            return 1;
        }
    };

    // Warn if report is stale (>120 min)
    if let Ok(metadata) = std::fs::metadata(&report_path) {
        if let Ok(modified) = metadata.modified() {
            if let Ok(elapsed) = modified.elapsed() {
                let age_min = elapsed.as_secs() / 60;
                if age_min > 120 {
                    eprintln!(
                        "[check_done] WARNING: Test report is {} min old — results may be stale; re-run tests for accurate check",
                        age_min
                    );
                }
            }
        }
    }

    let report = match prd::read_json(&report_path) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("[check_done] ERROR: {}", e);
            return 1;
        }
    };

    let summary = report.get("summary").cloned().unwrap_or_default();
    let failed = summary.get("failed").and_then(|v| v.as_i64()).unwrap_or(0);
    let errored = summary.get("errored").and_then(|v| v.as_i64()).unwrap_or(0);
    let passed = summary.get("passed").and_then(|v| v.as_i64()).unwrap_or(0);
    let test_total = summary.get("total").and_then(|v| v.as_i64()).unwrap_or(0);
    let pass_rate = summary
        .get("pass_rate")
        .and_then(|v| v.as_str())
        .unwrap_or("?");

    let report_name = Path::new(&report_path)
        .parent()
        .and_then(|p| p.file_name())
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_default();

    println!(
        "[check_done] Tests ({}): {}/{} pass ({}), {} failed, {} errored",
        report_name, passed, test_total, pass_rate, failed, errored
    );

    let prd_done = pending.is_empty();
    let tests_clean = failed == 0 && errored == 0;

    if prd_done && tests_clean {
        println!("[check_done] RESULT: SPIRAL COMPLETE — all stories done and 100% tests pass!");
        return 0;
    }

    let mut reasons = Vec::new();
    if !prd_done {
        reasons.push(format!("{} pending stories", pending.len()));
    }
    if !tests_clean {
        reasons.push(format!("{} test failure(s)", failed + errored));
    }
    println!(
        "[check_done] RESULT: INCOMPLETE ({})",
        reasons.join(", ")
    );
    1
}
